File: apps/api/app/schemas/catalog.py
import uuid
from datetime import datetime, time

from pydantic import BaseModel, Field

# ── City ──────────────────────────────────────────────────────────────────────

class CityRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    name: str
    slug: str
    timezone: str
    is_active: bool


class CityCreate(BaseModel):
    name: str = Field(max_length=100)
    slug: str = Field(max_length=100, pattern=r"^[a-z0-9-]+$")
    timezone: str = Field(max_length=64)


class CityUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=100)
    timezone: str | None = Field(default=None, max_length=64)
    is_active: bool | None = None


# ── Vendor ────────────────────────────────────────────────────────────────────

class VendorRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    city_id: uuid.UUID
    owner_user_id: uuid.UUID | None
    name: str
    slug: str
    description: str | None
    logo_url: str | None
    is_active: bool


class VendorCreate(BaseModel):
    city_id: uuid.UUID
    name: str = Field(max_length=255)
    slug: str = Field(max_length=255, pattern=r"^[a-z0-9-]+$")
    description: str | None = None
    logo_url: str | None = None


class VendorUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    description: str | None = None
    logo_url: str | None = None
    is_active: bool | None = None


# ── Category ──────────────────────────────────────────────────────────────────

class CategoryRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    parent_id: uuid.UUID | None
    name: str
    slug: str
    sort_order: int
    is_active: bool
    children: list["CategoryRead"] = []


class CategoryCreate(BaseModel):
    parent_id: uuid.UUID | None = None
    name: str = Field(max_length=100)
    slug: str = Field(max_length=100, pattern=r"^[a-z0-9-]+$")
    sort_order: int = 0


class CategoryUpdate(BaseModel):
    parent_id: uuid.UUID | None = None
    name: str | None = Field(default=None, max_length=100)
    sort_order: int | None = None
    is_active: bool | None = None


# ── ProductImage ──────────────────────────────────────────────────────────────

class ProductImageRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    url: str
    alt_text: str | None
    sort_order: int
    is_primary: bool


# ── ProductCity ───────────────────────────────────────────────────────────────

class ProductCityRead(BaseModel):
    model_config = {"from_attributes": True}

    city_id: uuid.UUID
    price_override_pkr: int | None
    delivery_fee_pkr: int
    lead_time_hours: int
    same_day_cutoff: time | None
    is_available: bool


class ProductCityUpsert(BaseModel):
    city_id: uuid.UUID
    price_override_pkr: int | None = Field(default=None, ge=0)
    delivery_fee_pkr: int = Field(default=0, ge=0)
    lead_time_hours: int = Field(default=24, ge=0)
    same_day_cutoff: time | None = None
    is_available: bool = True


# ── ProductVariant ────────────────────────────────────────────────────────────

class ProductVariantRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    name: str
    price_delta_pkr: int
    stock_qty: int
    attrs: dict[str, object] | None
    is_active: bool


# ── VendorSummary (embedded in ProductRead) ───────────────────────────────────

class VendorSummary(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    name: str
    slug: str


# ── Product ───────────────────────────────────────────────────────────────────

class ProductRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    vendor_id: uuid.UUID
    category_id: uuid.UUID | None
    name: str
    slug: str
    description: str | None
    base_price_pkr: int
    is_active: bool
    images: list[ProductImageRead] = []
    product_cities: list[ProductCityRead] = []
    variants: list[ProductVariantRead] = []
    vendor: VendorSummary | None = None


class ProductCreate(BaseModel):
    vendor_id: uuid.UUID
    category_id: uuid.UUID | None = None
    name: str = Field(max_length=255)
    slug: str = Field(max_length=255, pattern=r"^[a-z0-9-]+$")
    description: str | None = None
    base_price_pkr: int = Field(ge=0)


class ProductUpdate(BaseModel):
    category_id: uuid.UUID | None = None
    name: str | None = Field(default=None, max_length=255)
    description: str | None = None
    base_price_pkr: int | None = Field(default=None, ge=0)
    is_active: bool | None = None


# ── Occasion ──────────────────────────────────────────────────────────────────

class OccasionRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    slug: str
    name: str
    name_ur: str | None
    banner_url: str | None
    starts_at: datetime | None
    ends_at: datetime | None
    sort_order: int
    is_active: bool


class OccasionCreate(BaseModel):
    slug: str = Field(max_length=100, pattern=r"^[a-z0-9-]+$")
    name: str = Field(max_length=100)
    name_ur: str | None = Field(default=None, max_length=100)
    banner_url: str | None = None
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    sort_order: int = 0


class OccasionUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=100)
    name_ur: str | None = Field(default=None, max_length=100)
    banner_url: str | None = None
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    sort_order: int | None = None
    is_active: bool | None = None


# ── List responses ────────────────────────────────────────────────────────────

class PaginatedProducts(BaseModel):
    items: list[ProductRead]
    total: int
    page: int
    page_size: int

File: apps/web/lib/catalog.ts
// Server-safe catalog fetch helpers for Next.js server components.
// Uses NEXT_PUBLIC_API_URL so it works in both server and client contexts in dev.
const API_BASE =
  process.env.API_INTERNAL_URL ??
  process.env.NEXT_PUBLIC_API_URL ??
  "http://localhost:8000/api/v1";

export type ProductImage = {
  id: string;
  url: string;
  alt_text: string | null;
  sort_order: number;
  is_primary: boolean;
};

export type ProductCity = {
  city_id: string;
  price_override_pkr: number | null;
  delivery_fee_pkr: number;
  lead_time_hours: number;
  same_day_cutoff: string | null;
  is_available: boolean;
};

export type ProductVariant = {
  id: string;
  name: string;
  price_delta_pkr: number;
  stock_qty: number;
  attrs: Record<string, unknown> | null;
  is_active: boolean;
};

export type VendorSummary = {
  id: string;
  name: string;
  slug: string;
};

export type Product = {
  id: string;
  vendor_id: string;
  category_id: string | null;
  name: string;
  slug: string;
  description: string | null;
  base_price_pkr: number;
  is_active: boolean;
  images: ProductImage[];
  product_cities: ProductCity[];
  variants: ProductVariant[];
  vendor: VendorSummary | null;
};

export type PaginatedProducts = {
  items: Product[];
  total: number;
  page: number;
  page_size: number;
};

export type Category = {
  id: string;
  parent_id: string | null;
  name: string;
  slug: string;
  sort_order: number;
  is_active: boolean;
  children: Category[];
};

export type Vendor = {
  id: string;
  city_id: string;
  name: string;
  slug: string;
  description: string | null;
  logo_url: string | null;
  is_active: boolean;
};

export type City = {
  id: string;
  name: string;
  slug: string;
  timezone: string;
  is_active: boolean;
};

export type Occasion = {
  id: string;
  slug: string;
  name: string;
  name_ur: string | null;
  banner_url: string | null;
  starts_at: string | null;
  ends_at: string | null;
  sort_order: number;
  is_active: boolean;
};

async function apiFetch<T>(path: string): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, { next: { revalidate: 60 } });
  if (!res.ok) throw new Error(`API ${res.status} — ${path}`);
  return res.json() as Promise<T>;
}

export const catalogApi = {
  getCategories: () => apiFetch<Category[]>("/categories"),

  getCities: () => apiFetch<City[]>("/cities"),

  getProducts: (params?: {
    category_id?: string;
    vendor_id?: string;
    city_id?: string;
    occasion_id?: string;
    page?: number;
    page_size?: number;
  }) => {
    const qs = new URLSearchParams();
    if (params?.category_id) qs.set("category_id", params.category_id);
    if (params?.vendor_id) qs.set("vendor_id", params.vendor_id);
    if (params?.city_id) qs.set("city_id", params.city_id);
    if (params?.occasion_id) qs.set("occasion_id", params.occasion_id);
    if (params?.page) qs.set("page", String(params.page));
    if (params?.page_size) qs.set("page_size", String(params.page_size));
    const query = qs.toString();
    return apiFetch<PaginatedProducts>(`/products${query ? `?${query}` : ""}`);
  },

  getProduct: (slug: string) => apiFetch<Product>(`/products/${slug}`),

  getVendor: (slug: string) => apiFetch<Vendor>(`/vendors/${slug}`),

  getOccasions: () => apiFetch<Occasion[]>("/occasions"),

  getOccasion: (slug: string) => apiFetch<Occasion>(`/occasions/${slug}`),

  getCategoryBySlug: async (slug: string): Promise<Category | null> => {
    const all = await catalogApi.getCategories();
    const flat = (cats: Category[]): Category[] =>
      cats.flatMap((c) => [c, ...flat(c.children)]);
    return flat(all).find((c) => c.slug === slug) ?? null;
  },
};

export function primaryImage(product: Product): string {
  const primary = product.images.find((i) => i.is_primary);
  return (
    primary?.url ??
    product.images[0]?.url ??
    "https://images.unsplash.com/photo-1513104890138-7c749659a591?w=800"
  );
}
